New Assisted Tramming (and Wizard) + accompanying changes and fixes

lukasradek opened the pull request:
### Summary
Adds completely reworked Tramming Wizard and unifies it with G35 tramming.
Tramming wizard supports from 3 to 9 (used to be 6) points, which is handy for large beds with 3x3 screw grids.
Has more practical and intuitive menu system, supports milimeters and "screw turn" units.

Menu (loop) patch to handle STATIC_ITEM in other places then just "one on top as a menu header".

Adds screen that can show custom long (multiline with autowrapping) message.

Language changes to reflect all mentioned above.
Some small fixes.

### Description
Full description on per file basis of changes and reasoning behind them.

**Marlin/src/lcd/menu/menu_tramming.cpp**
This is the Tramming Wizard (TW) implementation.
Tramming wizard was a great idea but it was underbaked a little bit. It was functional but far from comfortably usable.
New tramming wizard workflow and possibilities:
1. **User opens TW and is prompted to select point to measure the reference Z value or select `Z=0` for absolute measurement.** The **value** is remembered, not the point! That is important and different from previous version.
**BEFORE:** When you adjusted the whole bed and decided to remeasure the reference **point**, your reference **height** would automatically change (without notifying user), which would throw off all measurements you made on already adjusted screws and the reference point would still show zero. If you wanted to change reference point, you had to close and open TW again and loose all measurements.
**NOW:** The reference **height** never changes unless user _intentionally_ chooses to. Even if user measures the point he used for reference again, the reference height stays the same, therefore the reading will show how the height at that point changed.
At anytime user is able to change the reference height just by selecting a point from which the height should be taken (or Z=0). No measurements are lost, they are just updated to reflect new reference height.
2.  **After the reference height is selected. _Measuring menu_ is presented to the user.**
There is a list of all _tramming point menu items_.
**BEFORE:** User had to first select a point, which would take him to _point submenu_. There he would have the option to measure a value or go back.
**NOW:** User is able to measure each point just by clicking the corresponding menu item. Measured value is displayed directly in that menu item's title, aligned to the right. (That is why this PR needs [that PR to be accepted](https://github.com/MarlinFirmware/Marlin/pull/25619)). There is no need to go into submenus for measurement.
4. **In the _Measuring menu_ there are further options.** (None of which were in the previous version).
User can **Measure All** points at once (the printer does it sequentially of course 😉)
User can **Toggle between units** - **Milimeters** / **Screw turns** (if `TRAMMING_SCREW_THREAD` was defined in config).
User can **Toggle menu mode - Measuring / Point detail mode**
User can **Invoke Reference height measurement** again
5. **Point detail mode**
I have already described _Measuring mode_ above. That is where you measure point by clicking menu items.
When user selects _Point detail mode_, all _tramming point menu items_ change from `ACTION_ITEM` to `SUBMENU_ITEM`. Now clicking on the item takes user into submenu for the specific tramming point.
In **tramming point submenu (point detail)** user is presented with height difference in _milimeters_ and _screw turns_. There is an option to _measure_ and _set this point's height as reference_ height.
6. Logic related change: **I have switched the way milimeter difference is presented.** Now it shows height differences in a more logical (cartesian) way. If point shows +1mm, it means it is 1mm above reference and vice versa. (It used to be the other way around). 

**Marlin\src\gcode\bedlevel\G35.cpp**
This is the implementation for G35 GCODE tramming.
I have made some changes to improve the reported information.
I have fixed a mistake... it used to output "Turn <point> clockwise by 1 turn and 30 minutes".
1 full turn is 360 degrees, 1 degree is 60 minutes.
Now the same instruction would output "... 1 turn and 180 degrees".

**Marlin\Configuration_adv.h**
1. **Remove `REPORT_TRAMMING_MM`**. Milimeters are universal and do not require any additional information (as opposed to screw turns, which require correct thread info). My tramming wizard and G35 show milimeters by default and screw turns only if user explicitly defines screw thread.
2. **I have made it more clear what CW and CCW means.** Yes... most of us know that (C)CW means (Counter)ClockWise, but what does it mean in the bed tramming context? Depending on the bed screw mechanism, tightening the screw can sometimes lower the bed, other time raise it. So _now_ user is clearly instructed that CW/CCW is the direction of bed lowering. No more ambiguity.

**Marlin\src\feature\tramming.h
Marlin\src\feature\tramming.cpp
Marlin\src\core\macros.h**
Those were modified to increase the maximum number of points from 6 to 9.
The idea was mine, but the Macro Magician who made it look nice and concise was @thinkyhead . Thanks.

**Marlin\src\lcd\language\language_xx.h**
Added some strings to show in the wizard and removed some from the old wizard.

**Marlin\src\lcd\marlinui.h
Marlin\src\lcd\menu\menu.cpp**
1. **There was a bug in MarlinUI::synchronize(...).**
Line 238: static FSTR_P sync_message = fmsg ?: GET_TEXT_F(MSG_MOVING);
Since this was a static initialization, it was evaluated only the first time this function ran. Therefore any subsequent call showed the same message as the first one (or garbage). Noone probably run into this issue since it is not that frequent or important.
But now it is fixed and I tied it nicely into my other UI function (which is new).
2. **MarlinUI::goto_message_screen(str1, str2)**
I needed / wanted a way to inform user by a full screen message (which also conveniently prevents further input).
I also wanted some flexibility, so this function can accept up to 2 string pointers, join them seemlessly and print them while wrapping the text to the screen. User (of this function) can also define linebreaks by inserting `\n` in the string.
It has some limitations - prints from `LCD_HEIGHT / 2 - 1` up to the end of the screen and does not give a damn about words (when wrapping), but well... it works for a reasonable input.
**While it would be a _nice to have_ for my Tramming Wizard, it is not neccessary if it is deemed not worthy. It can be easily dropped from wizard's code.**

**Marlin\src\lcd\menu\menu_item.h**
**Crucial bug fix:**
Original code considered `STATIC_ITEM`s to be used in _menu screens_ only as a single line headline (at the very top).
That raised some problems when I wanted to use `STATIC_ITEM` as a separator in menu and as a double line headline.
The problem was that `STATIC_ITEMS` were able to jump your encoderLine ("selection cursor") underneath them (`_skipStatic`). That is so that when you open menu with `STATIC_ITEM` at the top, your encoderLine is on the first clickable item and not the `STATIC_ITE`M. That is fine... but `STATIC_ITEM` does not care if you are coming from above them or from the bottom.. they always skip you down. Meaning that if you found yourself under `STATIC_ITEM` it would not let you scroll above it even though it wasn't the top item in the menu. Very... selfish.
Line351 `&& encoderTopLine == 0 ` ... this condition prevents skipping unless you are scrolled all the way up in the menu.

**Less crutial _(requires quite special circumstances)_, but still a bug fix:**
There was also a bug, when multiple `STATIC_ITEM`s in a row (or column to be exact) `skipStatic`ed you to a line that was behind `LCD_HEIGHT`. In such case, the UI slept a bit on scrolling and it caused some problems with printing characters (it would print a few lines over themselves).
Line352 `&& encoderLine < encoderTopLine+LCD_HEIGHT-1)` ... this prevents skipping if it would skip you behind LCD_HEIGHT.


### Benefits

1. Much more user friendly and capable Tramming Wizard and a bit more user friendly G35 (it wasn't that bad before and I just matched it's logic to TW).
2. Clearer Assisted Tramming setup in Configuration_adv.h
3. Assisted Tramming is now useful for 3x3 screw grid beds.
4. Ability to display custom long message which blocks user input.
5. Fixed stuck synchronization message.
6. Static items can now be used in interactive menus at any position.
7. Me being happy about my first Marlin contribution 😉.

### Configurations

Assisted tramming and Wizard turned on.

### Related Issues

[[FR] More optimized ASSISTED_TRAMMING #24448](https://github.com/MarlinFirmware/Marlin/issues/24448)

### You can read all of the above or just look at pictures
(Sorry, I forgot to switch my -V flag off 😁)

Measure Mode vs Point Detail Mode & Milimeters vs Screw Turns
<p float="left">
  <img src="https://user-images.githubusercontent.com/16482224/216788340-8b184d34-a9b9-4d58-a217-b2c4ffff0f30.png" width="30%" />
  <img src="https://user-images.githubusercontent.com/16482224/216788401-ba965346-5c97-4302-8a2c-a7b5065be580.png" width="30%" /> 
</p>

Point detail menu & Multiline message screen
<p float="left">
  <img src="https://user-images.githubusercontent.com/16482224/216788940-b23f0ae1-5837-4d8c-a694-b0595fdb2541.png" width="30%" />
<img src="https://user-images.githubusercontent.com/16482224/216788962-987c69ca-c038-4c23-8d71-8af943a7fb5a.png" width="30%" />
</p>

lukasradek committed:
Menu Items - Right aligned string & Icons for TFT

Adds ability for generic and static menu items to show (in addition to their title) right aligned string. Visually similar to Edit items but non editable and can be used with ACTION, SUBMENU etc.
I have implemented and tested it for DOGM, HD44780 and TFT lcds.
I have implemented a simple fallback for e3v2 and TFTGLCD, but I have no way to test it.

Shows helpful menu icons on TFT_COLOR_UI in place of post_char. Shows the equivalent icons to DOGM display. ">" is displayed as "right arrow icon", LCD_STR_UPLEVEL is displayed as "back icon".

FUNCTION lcd_uint_t lcd_put_u8str_P(PGM_P const ptpl, const int8_t ind, const char *cstr/*=nullptr*/, FSTR_P const fstr/*=nullptr*/, const lcd_uint_t maxlen/*=LCD_WIDTH*/) in /lcd/lcdprint.cpp has fixed return value. It should return number of outputed characters, but it used to return number of remaining characters from given maxlen limit. Codebase is adjusted to this change, where it was neccessary.

rhapsodyv committed:
🧑‍💻 General 'MediaFile' type alias (#24424)

lukasradek committed:
New Assisted Tramming (and Wizard), Small menu fixes, Screen with custom message

Adds completely reworked Tramming Wizard and unifies it with G35 tramming.
Tramming wizard supports from 3 to 9 (used to be 6) points, which is handy for large beds with 3x3 screw grids.
Has more practical and intuitive menu system, supports milimeters and "screw turn" units.

Menu (loop) patch to handle STATIC_ITEM in other places then just "one on top as a menu header".

Adds screen that can show custom long (multiline with autowrapping) message.

Language changes to reflect all mentioned above.

There is also a Configuration.h and Configuration_adv.h change for demonstration purposes.

Co-Authored-By: Scott Lahteine <gitware@thinkyhead.com>

sbasmanov commented:
Fantastic! I'll test it later today when get home. Thank You so much.

lukasradek commented:
@sbasmanov Crossing fingers you printer doesn't blow up 😁. Btw what screen do you have? There is a need for testing TFTGLCD and E3V2.

lukasradek committed:
Option to hide post_char icons on TFT_COLOR_UI

It is controlled by TFT_UI_HIDE_CLASSIC_MARKS that can be defined in config.

sbasmanov commented:
> @sbasmanov Crossing fingers you printer doesn't blow up 😁. Btw what screen do you have? There is a need for testing TFTGLCD and E3V2.

Everything works perfect! Leveling never been that easy :) I have Ender 3 Pro with stock (CR10) display. 
![IMG_6072](https://user-images.githubusercontent.com/10542322/216794965-25188537-017f-4c1a-9c61-dbac3a65969c.jpg)
Couple notes:
1. When 'measure all' or moving between points, there is extra small lift of head. After measuring with bltouch it usually return to original height (probably it is somewhere in config, don't remember). So, probe goes down, up, then little more up, then to next point. Not sure if this 'extra up' can be configured somewhere. 
2. Not sure if this is intended, but when cursor reach menu item with dashes (at bottom), cursor disappears at all. But UI thinks this is menu item and don't skip it. So, extra wheel click needed to see cursor again. Probably not an issue on touch screens :)
![IMG_6074](https://user-images.githubusercontent.com/10542322/216794981-0348db8c-0566-4fc4-a0cb-2b848a203dcb.jpg)
3. May be make config option to hide initial 'select reference point' menu? Something like DEFAULT_REFERENCE_POINT and if defined, don't show selection when wizard started (it's still possible to change ref point in menu later).

So, now I can say I'm happy :) Thank You very much. Hope this PR will be accepted soon and become part of marlin.

Oh, btw, saw one warning:
Marlin/src/lcd/menu/menu_tramming.cpp:69:14: warning: 'float _mock_probe_at_point(xy_pos_t)' defined but not used [-Wunused-function]
   69 | static float _mock_probe_at_point(const xy_pos_t pos) {

lukasradek commented:
So nice to see it working somewhere else in the world 🙂.

1. I think that "extra up" is configurable. I think it is to protect the BLtouch. I haven't touched the motion code at all, so it is the same as it was before.
2. Selection disappears on the dashed line to indicate, that it is not an action item. The same goes for the top items in point detail menu. I am 50:50 on whether it should be indicated as selected or not.
3. I think there is no reason to hide "select reference point" menu. You have to make the first measurement anyway, so the menu instructs you to do it so that there is no confusion about what is the reference.
4. The warning is a non issue.

lukasradek committed:
Easily switch between simulator and real world tram test

It also takes care of the unused function warning.

sbasmanov commented:
>I think it is to protect the BLtouch. I haven't touched the motion code at all, so it is the same as it was before.

Hmmm. I'll check configuration. On 2.1.2 release with old wizard it was moving straight to next point without this hop. May be something changed after release - I applied diff to config files with my changes, so may be missed some defaults.

>I am 50:50 on whether it should be indicated as selected or not.

I think that non-action items should be skipped and cursor moved to next available action item for better user experience.

>You have to make the first measurement anyway, so the menu instructs you to do it so that there is no confusion about what is the reference.

Agree. Except situation when Z=0 :) If it will be configurable (#define in config) it could do automatic measurement of pre-configured point on wizard start. May be someone always using same corner for leveling.

Anyway, this is not an issue, because wizard itself works very well.

sbasmanov commented:
>I think that "extra up" is configurable. I think it is to protect the BLtouch.

Ok, I see what's going on. Before, I had to exit menu, select point, enter menu, then measure. This is why I didn't notice that extra small hop between points. Now when it does 'measure all' it is noticeable. Especially because Z speed is different when doing this extra hop.
So, what I found so far. After homing, it sets Z=10 + Zoffset. After probe_at_point it sets Z=5+Zoffset+measured height. 
It seems this behavior is controlled by 4 params:
#define Z_CLEARANCE_DEPLOY_PROBE   10 // Z Clearance for Deploy/Stow
#define Z_CLEARANCE_BETWEEN_PROBES  5 // Z Clearance between probe points
#define Z_CLEARANCE_MULTI_PROBE     5 // Z Clearance between multiple probes
//#define Z_AFTER_PROBING           5 // Z position after probing is done

Not sure if it's possible to return head to original Z, so all moves between points will be smooth. Also, may be I have different speeds for travel move and for probe move, so this is why I see that extra hop.

lukasradek committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into Assisted-Tramming-Reloaded

lukasradek committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into Menu-Items-with-Right-Aligned-Strings

lukasradek committed:
Change configuration token name

lukasradek committed:
Revert "🧑‍💻 General 'MediaFile' type alias (#24424)"

This reverts commit 7467f44b47232269c946ff4b9e6f4b374096a765.

lukasradek committed:
Remove empty lines

lukasradek committed:
Refactor

Remove duplicate REPLIST macros.
Remove misplaced bracket in language_en

lukasradek committed:
Remove files used for dev testing

lukasradek committed:
Update to reflect changes on bugfix

Specificaly merging this PR https://github.com/MarlinFirmware/Marlin/pull/25498

lukasradek committed:
Merge remote-tracking branch 'upstream/bugfix-2.1.x' into Assisted-Tramming-Reloaded-PR

thinkyhead commented:
The last few months have been focused on Input Shaping, Fixed-Time Motion, fixing a Linear Advance issue that appeared in 2.1.2, cleaning up prpbe behavior, and doing a major refactor of LCD pins … so I'm only now getting a chance to review feature additions and improvements like this one.

So, this PR also needs a `git merge bugfix-2.1.x` and reconciliation with the new `SS_FULL` style for aligning Label / Value items. As with #25619, I recommend reapplying this feature on top of the latest code rather than attempting the `git merge` directly. Then you can force-push the final result to this PR's branch to update the PR.

As there is some overlap with #25619, you might want to just combine those two PRs into a single PR.

thinkyhead committed:
cleanup

thinkyhead committed:
cleanup, partial fix

thinkyhead committed:
Replace '=' with '{' for substitution

thinkyhead committed:
Can use '='

thinkyhead commented on Marlin/src/lcd/menu/menu.cpp:
```diff
@@ -228,21 +229,73 @@ void MarlinUI::goto_screen(screenFunc_t screen, const uint16_t encoder/*=0*/, co
+            item_line_buf[item_line_pos] = '\0';
+            const uint8_t lcd_line = LCD_HEIGHT / 2 - 1 + line_to_draw;
+            if (lcd_line < LCD_HEIGHT) { // implementation limited
+              MenuItem_static::draw(lcd_line, FPSTR(item_line_buf));
```
It's not possible to coerce the SRAM string `item_line_buf` into a PROGMEM string. This will not work on AVR, and will just display garbage.

thinkyhead replied:
This message screen concept is basically the same (with no buttons) as `MenuItem_confirm::draw_select_screen` which uses the existing method `MarlinUI::draw_select_screen_prompt` to draw the provided `FSTR_P`. It uses `wrap_string_P(col, row, FTOP(fstr), true)` to split up PROGMEM strings that are wider than the screen and properly handles UTF-8.

lukasradek replied:
I fixed the SRAM/PROGMEM issue by using the `MenuItem_static::draw` with `nullptr` for `fstr` and `item_line_buf` as `vstr`.

Using `draw_select_screen` is a great idea with one caveat 🙂, it is not implemented for HAS_GRAPHICAL_TFT, because that uses completely separate select screen implementation. So for now I will be using my function so that the Tramming Wizard is done and I will consolidate those methods in the future.

thinkyhead committed:
fix some serial echo

thinkyhead committed:
clarify r_offset modification

thinkyhead committed:
misc. related updates

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/25619

thinkyhead committed:
cleanup

lukasradek commented:
Great Scott!! What a commit hurricane 😀. Thanks for looking into this @thinkyhead, you focus on technical features is well understood 🙂. Right now I am quite busy with other stuff, so it might take me a while to go throught this - I assume a few weeks.

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/25619

lukasradek committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into Menu-Items-with-Right-Aligned-Strings

lukasradek committed:
Merge remote-tracking branch 'upstream/bugfix-2.1.x' into Menu-Items-with-Right-Aligned-Strings

sbasmanov commented:
@lukasradek, is there any chance You will have time in near future to take a look at this? I'm using Your patch since day one, and I can say with 100% confidence, that this wizard is best leveling tool I ever tried. There are some new features and bug fixes in Marlin since April, but I won't move until Your patch will be available :)

lukasradek commented:
Hello @sbasmanov , that is very nice to hear!

Unfortunately when I had time, Marlin devs were busy with other features and now when they responded, I am busy with my non-Marlin life 🙂.
I definitely want to finish this though and I believe I will have more time to do that during July.

sbasmanov commented:
@lukasradek , that's normal :) All of us have our own life and it is not always aligns with our hobby :)
I just wanted to make sure You didn't forget about such great feature and some day this PR will be accepted. We're not in hurry, so July sounds great. Thank You.

lukasradek commented:
@sbasmanov  I think it might be 🙂. Especially since you praise it so much 😁. Btw... a little sneak peak of a future version.
![Snímek obrazovky_20230206_170628](https://github.com/MarlinFirmware/Marlin/assets/16482224/36373d8f-105c-48a9-9e74-48fd82c9c756)

lukasradek committed:
Merge remote-tracking branch 'upstream/bugfix-2.1.x' into Menu-Items-with-Right-Aligned-Strings

lukasradek committed:
Unify _ITEMs' formatting

(Almost) all _ITEMs now support aligning styles and value strings.

lukasradek committed:
Merge remote-tracking branch 'upstream/bugfix-2.1.x' into Assisted-Tramming-Reloaded-PR

lukasradek committed:
Merge branch 'Menu-Items-with-Right-Aligned-Strings' into Assisted-Tramming-Reloaded-PR

lukasradek committed:
Ported to new MENU_ITEM formatting

and fixed some bugs that were introduced in cleanup

lukasradek committed:
Fix RAM/ROM reading issue

lukasradek committed:
Rework of methods drawing Message Screens

Separate (naming-wise) method that draws messaage on screen from method that draws select prompt.

Implement method to draw message screen with wordwrap for TFT_COLOR_UI.

Remove code duplication from TFT method that draws select screens across different resolutions.

lukasradek commented:
I have separated (naming-wise) method that draws messaage on screen `MarlinUI::draw_message_on_screen` from method that draws select prompt.
Now developer can use that method to just draw a message screen (without select prompt) and select prompts now use that method to draw its message without code duplication.
And I have implemented that method for TFT_COLOR_UI as well since it previously used completely different logic to draw select screens.

My `MarlinUI::goto_message_screen` now uses that unified method `ui.draw_message_on_screen` and has the option to `synchronize` itself with planner. Meaning it should be possible to remove the separate `MarlinUI::sychronize` method which could handle only one string that would be force printed on one line regardless of the fact if it fits.

I have also removed some code duplication because different TFT resolutions had the same implementation of draw_select_screen in each resolution file.

It still need some refactor and cleanup.

lukasradek committed:
Merge remote-tracking branch 'upstream/bugfix-2.1.x' into Assisted-Tramming-Reloaded-PR

lukasradek committed:
Post-merge fix

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/25339

thinkyhead committed:
followup

thinkyhead committed:
also

thinkyhead committed:
and…

vlsi commented on Marlin/src/lcd/tft/ui_common.cpp:
```diff
@@ -416,6 +456,78 @@ void MarlinUI::clear_lcd() {
+  p = &p[wrd_len];
+  bool eol = !p;
+  while (!eol) {
+    wrd_len = _get_word((const char *)p, cb_read_byte, last_char);
+    const uint8_t len = tft_string.length;
+    tft_string.add((T)p, wrd_len);
+    const uint32_t wid = tft_string.width();
+
+    if (wid > TFT_WIDTH) {
+      tft_string.truncate(len);
+      print_str();
+      tft_string.set((T)p, wrd_len);
+    }
+    if (last_char == '\n') {
+      tft_string.truncate(tft_string.length - 1);
+      print_str();
+    }
+
+    p = &p[wrd_len];
```
This crashes with UTF-8 characters. `_get_word` returns the number of UTF-8 characters, while `p = &p[wrd_len]` advances by the number of bytes.
I guess `p = &p[wrd_len];` should be replaced with something like "advance by wrd_len utf-8 chars, or just return the next pointer from `_get_word`

lukasradek replied:
Yes, something like that will be neccessary.

vlsi reviewed (changes requested):
Thanks for implementing this. It looks great, however, I have some suggestions.

First, it crashes in Russian language, and I narrowed down the crash to misuse of `_get_word`.

Second, I wonder if you considered an alternative option for adjusting the screws.
What if the probe moves over the screw at exactly the height the probe should trigger.
Then user could adjust the screw until the probe triggers. Unfortunately, BLTouch would need to re-deploy after triggering, so it would probably require pressing "hover probe again" so user could tune the screw again.

vlsi commented on Marlin/src/lcd/marlinui.cpp:
```diff
@@ -472,20 +474,21 @@ void MarlinUI::init() {
-          wrap_string(col, row, string);
-        }
-        if (suff) wrap_string_P(col, row, FTOP(suff));
+
+      void MarlinUI::draw_message_on_screen(FSTR_P const pref, const char * const string/*=nullptr*/, FSTR_P const suff/*=nullptr*/) {
+          const uint8_t plen = utf8_strlen(pref), slen = suff ? utf8_strlen(suff) : 0;
+          uint8_t col = 0, row = 0;
+          if (!string && plen + slen <= LCD_WIDTH) {
+            col = (LCD_WIDTH - plen - slen) / 2;
+            row = LCD_HEIGHT > 3 ? 1 : 0;
+          }
+          if (LCD_HEIGHT >= 8) row = LCD_HEIGHT / 2 - 2;
+          wrap_string_P(col, row, FTOP(pref), true);
+          if (string) {
+            if (col) { col = 0; row++; } // Move to the start of the next line
+            wrap_string(col, row, string);
+          }
+          if (suff) wrap_string_P(col, row, FTOP(suff));
```
WDYT of un-indenting it back to 2-space indent?

vlsi commented on Marlin/src/lcd/tft/ui_common.cpp:
```diff
@@ -416,6 +456,78 @@ void MarlinUI::clear_lcd() {
+  tft_string.set();
+
+  if (plen) _wrap_string<FSTR_P>(row, pref, read_byte_rom);
+  if (string) _wrap_string<const char * const>(row, string, read_byte_ram);
```
```suggestion
  if (strlen) _wrap_string<const char * const>(row, string, read_byte_ram);
```

vlsi commented on Marlin/src/lcd/dogm/marlinui_DOGM.cpp:
```diff
@@ -447,11 +447,35 @@ void MarlinUI::clear_lcd() { } // Automatically cleared by Picture Loop
+      }
+      else {
+
+        n -= lcd_put_u8str(ftpl, itemIndex, itemStringC, itemStringF, n);
+        n *= MENU_FONT_WIDTH;
+        while (n > 0) n -= lcd_put_u8str(F(" "));
```
Multiplying by `MENU_FONT_WIDTH` looks odd here.
If `lcd_put_u8str(` returns the number of pixels, then multiplication should not be needed.
However, if `lcd_put_u8str(` returns the number of chars, then the line 474 looks odd as it

vlsi replied:
Did you mean the following?

```c
        n *= MENU_FONT_WIDTH;
        n -= lcd_put_u8str(ftpl, itemIndex, itemStringC, itemStringF, n);
        while (n > 0) n -= lcd_put_u8str(F(" "));
```

lukasradek replied:
The oddity is not the "out of place" multiplication... if I remember correctly the oddity is that lcd_put_u8str is not a single method - it is an overloaded method that returns different based on which overload it is. The first one (with string substitution) returns number of chars and the second one (with just single arg) returns pixels.

So my code is correct and if we want to fix the semantics, it would mean digging into the mess of overloads with different return specs 🙂.

vlsi commented on Marlin/src/lcd/TFTGLCD/marlinui_TFTGLCD.cpp:
```diff
@@ -1008,12 +1008,42 @@ void MarlinUI::draw_status_screen() {
+
+    // Draw as much of the label as fits
+    if (plen) {
+      const int8_t expl = n;
+      n = lcd_put_u8str(fstr, itemIndex, itemStringC, itemStringF, n);
+      pad -= (expl - n - plen); // Reduce the padding
```
Could you explain the maths in the comment?
`n = lcd_put_u8str` looks suspicious. Should it be something like `n -= lcd_put_u8str...`?

lukasradek committed:
Update Marlin/src/lcd/tft/ui_common.cpp

Co-authored-by: Vladimir Sitnikov <sitnikov.vladimir@gmail.com>

vlsi commented on Marlin/src/lcd/tft/ui_common.cpp:
```diff
@@ -416,6 +456,78 @@ void MarlinUI::clear_lcd() {
+
+  tft_string.set();
+
+  if (plen) _wrap_string<FSTR_P>(row, pref, read_byte_rom);
+  if (strlen) _wrap_string<const char * const>(row, string, read_byte_ram);
+  if (slen) _wrap_string<FSTR_P>(row, suff, read_byte_rom, true);
```
I believe this `flush=true/false` is not working as intended.
For instance, if `slen==0` and the message fits within a single line, then the string will never be flushed, and the screen will be blank.

In other words, I would suggest adding a separate flush after all `_wrap_string` calls.

vlsi commented on Marlin/src/lcd/menu/menu_tramming_wizard.cpp:
```diff
@@ -40,64 +41,226 @@
+  // i.e. whether the menu is in Reference point selection mode or Measuring mode
+
+  STATIC_ITEM_F(reference_valid ? GET_TEXT_F(MSG_TW_SELECT_TRAM_POINT) : GET_TEXT_F(MSG_TW_SELECT_TRAM_REF));
+  if (reference_valid) ACTION_ITEM_F(F("Measure All"), _probe_all_points);
```
localize, plz

vlsi commented on Marlin/src/lcd/menu/menu_tramming_wizard.cpp:
```diff
@@ -40,64 +41,226 @@
+
+  STATIC_ITEM_F(FPSTR(pgm_read_ptr(&tramming_point_name[tram_target])), SS_FULL, _get_mms_str(tram_target));
+  #ifdef TRAMMING_SCREW_THREAD
+    STATIC_ITEM_N_F(screw_size, F("M= screw turns"), SS_FULL, _get_screw_turns_str(tram_target));
```
localize, plz

vlsi commented:
Here's an alternative implementation that uses a probe+buzzer so the user can have audio and visual feedback when tuning screws: https://github.com/MarlinFirmware/Marlin/blob/d56136f06cca075e801aabcff76bd207d4da349f/Marlin/src/lcd/menu/menu_bed_tramming.cpp#L270-L277

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.1.x' into pr/25339

thinkyhead committed:
merge followup - imperfect merge

thinkyhead commented:
The `lcd_put_u8str` methods were a bit more standardized recently so they behave more consistently across different display types. I'm sure I didn't merge everything quite perfectly with these string methods. Please fix up these methods starting from the state of `bugfix-2.1.x` and bringing in your changes in the proper manner to fit the current code.

thinkyhead committed:
fix extra string

lukasradek commented:
@vlsi  As @sbasmanov mentioned, leveling by adjustment according to probe height is either PROBE_MANUALLY or AUTO_BED_LEVELING_MANUAL (or similar) in config, but that is leveling... not tramming. Even though in this case it sort of blends together.

vlsi commented:
> BLtouch can't trigger, when bed goes down

That is indeed unfortunate, however, the wizard could at least assist when the bed should tune up.

The wizard could measure four corners and then walk them in such an order so the bed is only moved up.

Inductive sensor can trigger both ways.

lukasradek commented:
> An inductive sensor can trigger both ways.

This is not entirely true. Inductive sensor can change state going both ways, but the "let go" distance is greater than the trigger distance. So the only viable way of doing this is to only instruct to adjust upwards, which can be reliably measured. Furthermore there is no way you can adjust downwards if you use NOZZLE_AS_PROBE 😄.

Anyway, I am willing to implement something like this but first I would like to see this merged in this "basic" form. I have put quite a lot of work into this already.

vlsi commented:
> I would like to see this merged in this "basic" form

Do you think you could fix utf-8 issues first?
In the current form, the PR is completely unworkable in Russian locale: it both causes "blank messages" and triggers board resets caused by utf-8 runaway.

lukasradek commented:
Sure, I can. But I don't have time for this right now unfortunately. Hopefully December will do it.

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/25339

thinkyhead committed:
followup

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/25339

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/25339

thinkyhead committed:
merge followup

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/25339

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/25339

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.1.x' into pr/25339

thinkyhead committed:
cleanup

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/25339

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/25339
